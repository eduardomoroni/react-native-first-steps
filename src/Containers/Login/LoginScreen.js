// @flow

import React, { Component } from 'react'
import {
  View,
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
  Image
} from 'react-native'
import { connect } from 'react-redux'
import Styles from '../../Styles/LoginScreenStyle'
import { Images, Metrics } from '../../Styles/Themes'
import { loginUser } from '../../Redux/Actions'
import { Actions as NavigationActions } from 'react-native-router-flux'
import I18n from 'react-native-i18n'
import { Spinner } from '../Components'

type LoginScreenProps = {
  error: string,
  loading: boolean,
  attemptLogin: () => void
}

class LoginScreen extends Component {
  props: LoginScreenProps

  state: {
    email: string,
    password: string,
    visibleHeight: number,
    topLogo: {
      width: number
    }
  }

  constructor (props: LoginScreenProps) {
    super(props)
    this.state = {
      email: '',
      password: '',
      visibleHeight: Metrics.screenHeight,
      topLogo: { width: Metrics.screenWidth }
    }
  }

  componentWillReceiveProps (newProps) {
    this.forceUpdate()

    if (newProps.user !== null) {
      NavigationActions.pop()
    } else {
      this.state.password = ''
    }
  }

  handlePressLogin = () => {
    const { email, password } = this.state

    this.props.attemptLogin({ email, password })
  }

  handleChangeEmail = (text) => {
    this.setState({ email: text })
  }

  handleChangePassword = (text) => {
    this.setState({ password: text })
  }

  renderButtons () {
    const { loading } = this.props

    if (loading) {
      return (
        <View style={[Styles.loginRow]}>
          <Spinner />
        </View>
      )
    }

    return (
      <View style={[Styles.loginRow]}>
        <TouchableOpacity style={Styles.loginButtonWrapper} onPress={this.handlePressLogin}>
          <View style={Styles.loginButton}>
            <Text style={Styles.loginText}>{I18n.t('signIn')}</Text>
          </View>
        </TouchableOpacity>
        <TouchableOpacity style={Styles.loginButtonWrapper} onPress={NavigationActions.pop}>
          <View style={Styles.loginButton}>
            <Text style={Styles.loginText}>{I18n.t('cancel')}</Text>
          </View>
        </TouchableOpacity>
      </View>
    )
  }

  render () {
    const { email, password } = this.state
    const { loading } = this.props
    const editable = !loading
    const textInputStyle = editable ? Styles.textInput : Styles.textInputReadonly

    return (
      <ScrollView contentContainerStyle={{justifyContent: 'center'}} style={[Styles.container, {height: this.state.visibleHeight}]} keyboardShouldPersistTaps='always'>
        <Image source={Images.logo} style={[Styles.topLogo, this.state.topLogo]} />
        <View style={Styles.form}>
          <View style={Styles.row}>
            <Text style={Styles.rowLabel}>{I18n.t('email')}</Text>
            <TextInput
              ref='email'
              style={textInputStyle}
              value={email}
              editable={editable}
              keyboardType='default'
              returnKeyType='next'
              autoCapitalize='none'
              autoCorrect={false}
              onChangeText={this.handleChangeEmail}
              underlineColorAndroid='transparent'
              onSubmitEditing={() => this.refs.password.focus()}
              placeholder={I18n.t('email')} />
          </View>
          <View style={Styles.row}>
            <Text style={Styles.rowLabel}>{I18n.t('password')}</Text>
            <TextInput
              ref='password'
              style={textInputStyle}
              value={password}
              editable={editable}
              keyboardType='default'
              returnKeyType='go'
              autoCapitalize='none'
              autoCorrect={false}
              secureTextEntry
              onChangeText={this.handleChangePassword}
              underlineColorAndroid='transparent'
              onSubmitEditing={this.handlePressLogin}
              placeholder={I18n.t('password')} />
          </View>

          {this.renderButtons()}

          <View>
            <Text style={Styles.rowLabel}>{this.props.error}</Text>
          </View>
        </View>

      </ScrollView>
    )
  }
}

const mapStateToProps = (state) => {
  const { loading, error } = state.auth
  const { user } = state.user

  return {
    loading: loading,
    error: error,
    user: user
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    attemptLogin: (credentials) => dispatch(loginUser(credentials))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(LoginScreen)
