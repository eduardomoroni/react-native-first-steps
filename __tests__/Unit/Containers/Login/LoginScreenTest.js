// @flow

import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import { LoginScreen } from '../../../../src/Containers/Login/LoginScreen'

const props = {
  loading: false,
  token: undefined,
  error: '',
  facebookLogin: jest.fn(),
  facebookLogout: jest.fn()
}

describe('<LoginScreen />', () => {
  it('Snapshot WIP', () => {
    const wrapper = shallow(<LoginScreen {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
