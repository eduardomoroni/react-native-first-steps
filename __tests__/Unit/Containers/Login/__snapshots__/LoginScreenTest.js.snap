// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> Snapshot 1`] = `
<ScrollViewMock
  contentContainerStyle={
    Object {
      "justifyContent": "center",
    }
  }
  keyboardShouldPersistTaps="always"
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "paddingTop": 94,
      },
      Object {
        "height": 1334,
      },
    ]
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "alignSelf": "center",
          "resizeMode": "contain",
        },
        Object {
          "width": 750,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 4,
        "margin": 10,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 20,
          "paddingVertical": 20,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#595959",
          }
        }
      >
        I18n.t(email)
      </Text>
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        editable={true}
        keyboardType="default"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="I18n.t(email)"
        returnKeyType="next"
        style={
          Object {
            "color": "#2d2d2d",
            "height": 40,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 20,
          "paddingVertical": 20,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#595959",
          }
        }
      >
        I18n.t(password)
      </Text>
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        editable={true}
        keyboardType="default"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="I18n.t(password)"
        returnKeyType="go"
        secureTextEntry={true}
        style={
          Object {
            "color": "#2d2d2d",
            "height": 40,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "paddingBottom": 20,
            "paddingHorizontal": 20,
          },
        ]
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        focusedOpacity={0.7}
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#161616",
              "borderColor": "#595959",
              "borderWidth": 1,
              "flex": 1,
              "padding": 6,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#F7F7F7",
                "textAlign": "center",
              }
            }
          >
            I18n.t(signIn)
          </Text>
        </View>
      </TouchableOpacity>
      <TouchableOpacity
        activeOpacity={0.2}
        focusedOpacity={0.7}
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#161616",
              "borderColor": "#595959",
              "borderWidth": 1,
              "flex": 1,
              "padding": 6,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#F7F7F7",
                "textAlign": "center",
              }
            }
          >
            I18n.t(cancel)
          </Text>
        </View>
      </TouchableOpacity>
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#595959",
          }
        }
      />
    </View>
  </View>
</ScrollViewMock>
`;
